File: Queues/java/src/QueueWithTwoStacks.java
import java.util.Stack;

public class QueueWithTwoStacks {
    private final Stack<Integer> inputStack = new Stack<>();
    private final Stack<Integer> outputStack = new Stack<>();

    // O(1)
    public void enqueue(int item) {
        inputStack.push(item);
    }

    // O(n)
    public int dequeue() {
        moveStack1toStack2();

        return outputStack.pop();
    }

    private void moveStack1toStack2() {
        if (isEmpty())
            throw new IllegalStateException();

        if (outputStack.isEmpty())
            while (!inputStack.isEmpty())
                outputStack.push(inputStack.pop());
    }

    public boolean isEmpty() {
        return inputStack.isEmpty() && outputStack.isEmpty();
    }

    public int peek() {
        moveStack1toStack2();
        return outputStack.peek();
    }


}

File: Queues/cpp/QueueWithTwoStacks.cpp
#include "QueueWithTwoStacks.h"

bool QueueWithTwoStacks::isEmpty() {
    return inputStack.empty() && outputStack.empty();
}

void QueueWithTwoStacks::enqueue(int item) {
    inputStack.push(item);
}

int QueueWithTwoStacks::dequeue() {
    MoveinputStackTooutputStack();

    int ret = outputStack.top();
    outputStack.pop();
    return ret;
}

void QueueWithTwoStacks::MoveinputStackTooutputStack() {
    if (isEmpty())
        throw "Queue is empty";

    if (outputStack.empty())
        while (!inputStack.empty()) {
            outputStack.push(inputStack.top());
            inputStack.pop();
        }
}

int QueueWithTwoStacks::peek() {
    MoveinputStackTooutputStack();
    return outputStack.top();
}
